### ISSUE TITLE
Wait timeout

### ISSUE THREAD

**Issue by dmitry-tomp** [debonet/es6semaphore#1]
Would be good to accept optional timeout in millis in `wait` function, so that the application logic can gracefully fail if there is too much load or some operations are taking too long.

**Comment by debonet** [comment]
Hi Dmitry, you can use @debonet/within (https://github.com/debonet/within) to achieve that. e.g.:

```
  const within = require( '@debonet/within' );
  const Semaphore = require( '@debonet/es6semaphore' );

  const sem = new Semaphore( 5 );

  ...
  // this could take a very long time
  await sem.wait();

  // whereas this will timeout after 2 seconds (see @debonet/within documentation for how to detect a timeout)
  await within( 2000, () => sem.wait());

  ...
  // or, if you're going to do it a lot, you might functionalize it:
  const get_semaphore_or_timeout_after_1s = within.limit( 1000, sem.wait.bind( sem ));
  await get_semaphore_or_timeout_after_1s();
```

**Comment by dmitry-tomp** [comment]
Thanks @debonet for sharing a nice idea!
I guess the `within` function is not going to cancel the `wait`, so eventually `wait()` promise might resolve as if we had successfully claimed the resource, even though we did not. I believe this will eventually lead to a leak.
I could try solving this by adding some simple timeout handling logic, like this:
```
await within(2000, () => sem.wait()).timedout((err) => {
  sem.release();
  throw err;
});
```
I also see that this semaphore does implement a fair locking guarantee, hope it can be maintained in case of any updates.
Actually I have one more feature request to raise, let me create another issue for that.
